File: atak/ATAKMapEngine/lib/src/main/java/com/atakmap/map/layer/raster/mobac/CustomMultiLayerMobacMapSource.java

package com.atakmap.map.layer.raster.mobac;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.Arrays;

import com.atakmap.coremap.maps.coords.GeoBounds;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Paint;

public class CustomMultiLayerMobacMapSource implements MobacMapSource {

    private final String name;
    private final int srid;
    private final float[] layerAlpha;
    private final MobacMapSource[] layers;
    private final int minZoom;
    private final int maxZoom;
    private final int backgroundColor;
    private Config config = new Config();

    public CustomMultiLayerMobacMapSource(String name, MobacMapSource[] layers, float[] layerAlpha,
            int backgroundColor) {
        this.name = name;
        this.layers = layers;
        
        if (layerAlpha == null) {
            layerAlpha = new float[this.layers.length];
            Arrays.fill(layerAlpha, 1.0f);
        }
        this.layerAlpha = layerAlpha;

        int min = this.layers[0].getMinZoom();
        int max = this.layers[0].getMaxZoom();
        for (int i = 0; i < this.layers.length; i++) {
            if (this.layers[i].getMinZoom() < min)
                min = this.layers[i].getMinZoom();
            if (this.layers[i].getMaxZoom() > max)
                max = this.layers[i].getMaxZoom();
        }
        this.minZoom = min;
        this.maxZoom = max;
        this.backgroundColor = backgroundColor;
        
        this.srid = this.layers[0].getSRID();
        for(int i = 1; i < this.layers.length; i++) {
            if(this.layers[i].getSRID() != this.srid)
                throw new IllegalArgumentException();
        }
    }

    @Override
    public int getSRID() {
        return this.srid;
    }

    @Override
    public final void clearAuthFailed() {
        for (int i = 0; i < this.layers.length; i++)
            this.layers[i].clearAuthFailed();
    }

    @Override
    public final String getName() {
        return this.name;
    }

    @Override
    public final int getTileSize() {
        return this.layers[0].getTileSize();
    }

    @Override
    public final int getMinZoom() {
        return this.minZoom;
    }

    @Override
    public final int getMaxZoom() {
        return this.maxZoom;
    }

    @Override
    public final String getTileType() {
        return "JPG";
    }

    @Override
    public GeoBounds getBounds() {
        return null;
    }

    @Override
    public final void checkConnectivity() {
        for (int i = 0; i < this.layers.length; i++)
            this.layers[i].checkConnectivity();
    }

    @Override
    public MobacMapTile loadTile(int zoom, int x, int y, BitmapFactory.Options opts)
            throws IOException {
        long expiration = Long.MAX_VALUE;

        Bitmap composite = Bitmap.createBitmap(this.getTileSize(), this.getTileSize(),
                Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(composite);
        canvas.drawColor(this.backgroundColor);

        Paint alphaPaint = new Paint();

        BitmapFactory.Options tileOpts = null;
        if(opts != null && opts.inJustDecodeBounds) {
            tileOpts = new BitmapFactory.Options();
            tileOpts.inDensity = opts.inDensity;
            tileOpts.inDither = opts.inDither;
            tileOpts.inInputShareable = opts.inInputShareable;
            tileOpts.inJustDecodeBounds = false;
            tileOpts.inMutable = opts.inMutable;
            tileOpts.inPreferQualityOverSpeed = opts.inPreferQualityOverSpeed;
            tileOpts.inPreferredConfig = opts.inPreferredConfig;
            tileOpts.inPurgeable = opts.inPurgeable;
            tileOpts.inSampleSize = opts.inSampleSize;
            tileOpts.inScaled = opts.inScaled;
            tileOpts.inScreenDensity = opts.inScreenDensity;
            tileOpts.inTargetDensity = opts.inTargetDensity;
            tileOpts.inTempStorage = opts.inTempStorage;
            tileOpts.mCancel = opts.mCancel;
        }

        MobacMapTile tile;
        for (int i = 0; i < this.layers.length; i++) {
            if (zoom < this.layers[i].getMinZoom() || zoom > this.layers[i].getMaxZoom())
                continue;
            tile = null;
            try {
                tile = this.layers[i].loadTile(zoom, x, y, tileOpts);
                if (tile == null || tile.bitmap == null)
                    continue;
                if (tile.expiration < expiration)
                    expiration = tile.expiration;
                alphaPaint.setAlpha((int) (this.layerAlpha[i] * 255));
                canvas.drawBitmap(tile.bitmap, 0, 0, alphaPaint);
            } finally {
                if (tile != null && tile.bitmap != null)
                    tile.bitmap.recycle();
            }
        }

        byte[] data = null;

        ByteArrayOutputStream os = new ByteArrayOutputStream();
        try {
            if (composite.compress(Bitmap.CompressFormat.PNG, 80, os))
                data = os.toByteArray();
        } finally {
            os.close();
        }

        if(opts != null && opts.inJustDecodeBounds && composite != null) {
            opts.outWidth = composite.getWidth();
            opts.outHeight = composite.getHeight();

            composite.recycle();
            composite = null;
        }

        return new MobacMapTile(composite, data, expiration);
    }

    @Override
    public void setConfig(Config c) {
        config = c;
    }

}

File: takkernel/engine/src/main/cpp/raster/mobac/CustomMultiLayerMobacMapSource.cpp
#include "raster/mobac/CustomMultiLayerMobacMapSource.h"

#ifdef MSVC
#include <msclr/gcroot.h>
#endif

#include "raster/mobac/MobacMapTile.h"

using namespace atakmap::raster::mobac;

namespace
{
    int _getMinZoom(std::vector<std::pair<MobacMapSource *, float>> &layers)
    {
        std::vector<std::pair<MobacMapSource *, float>>::iterator it;

        it = layers.begin();
        int min = (*it).first->getMinZoom();
        for ( ; it != layers.end(); it++) {
            if ((*it).first->getMinZoom() < min)
                min = (*it).first->getMinZoom();
        }
        return min;
    }

    int _getMaxZoom(std::vector<std::pair<MobacMapSource *, float>> &layers)
    {
        std::vector<std::pair<MobacMapSource *, float>>::iterator it;

        it = layers.begin();
        int max = (*it).first->getMinZoom();
        for (; it != layers.end(); it++) {
            if ((*it).first->getMaxZoom() > max)
                max = (*it).first->getMaxZoom();
        }
        return max;
    }
}

CustomMultiLayerMobacMapSource::CustomMultiLayerMobacMapSource(const char *n, std::vector<std::pair<MobacMapSource *, float>> &l, int bg) :
    name(NULL),
    layers(l),
    minZoom(_getMinZoom(l)),
    maxZoom(_getMaxZoom(l)),
    backgroundColor(bg)
{
    char *str = new char[strlen(n) + 1];
    sprintf(str, "%s", n);
    name = str;
}

CustomMultiLayerMobacMapSource::~CustomMultiLayerMobacMapSource()
{
    std::vector<std::pair<MobacMapSource *, float>>::iterator it;
    for (it = layers.begin(); it != layers.end(); it++)
        delete (*it).first;
    layers.clear();
}

void CustomMultiLayerMobacMapSource::clearAuthFailed()
{
    std::vector<std::pair<MobacMapSource *, float>>::iterator it;
    for (it = layers.begin(); it != layers.end(); it++)
        (*it).first->clearAuthFailed();
}

const char *CustomMultiLayerMobacMapSource::getName()
{
    return this->name;
}

int CustomMultiLayerMobacMapSource::getTileSize()
{
    return this->layers[0].first->getTileSize();
}

int CustomMultiLayerMobacMapSource::getMinZoom()
{
    return this->minZoom;
}

int CustomMultiLayerMobacMapSource::getMaxZoom()
{
    return this->maxZoom;
}

const char *CustomMultiLayerMobacMapSource::getTileType()
{
    return "JPG";
}

bool CustomMultiLayerMobacMapSource::getBounds(atakmap::feature::Envelope *bnds)
{
    return false;
}

void CustomMultiLayerMobacMapSource::checkConnectivity()
{
    std::vector<std::pair<MobacMapSource *, float>>::iterator it;
    for (it = layers.begin(); it != layers.end(); it++)
        (*it).first->checkConnectivity();
}

bool CustomMultiLayerMobacMapSource::loadTile(MobacMapTile *tile, int zoom, int x, int y/*, BitmapFactory.Options opts*/) /*throws IOException*/
{
    // XXX - no means to composite via SDK Bitmap API

#ifdef MSVC
    int64_t expiration = INT64_MAX;

    System::Drawing::Bitmap ^composite = gcnew System::Drawing::Bitmap(this->getTileSize(),
                                                                       this->getTileSize());

    System::Drawing::Graphics ^canvas = System::Drawing::Graphics::FromImage(composite);
    if (this->backgroundColor != 0)
        canvas->Clear(System::Drawing::Color::FromArgb(this->backgroundColor));

    System::Drawing::Imaging::ImageAttributes alphaPaint;

    bool gotTile;
    
    std::vector<std::pair<MobacMapSource *, float>>::iterator it;
    for (it = layers.begin(); it != layers.end(); it++) {
        if (zoom < (*it).first->getMinZoom() || zoom >(*it).first->getMaxZoom())
            continue;

        MobacMapTile layerTile;
        try {
            gotTile = (*it).first->loadTile(&layerTile, zoom, x, y/*, opts*/);
            if (gotTile)
                continue;
            if (tile->expiration < expiration)
                expiration = tile->expiration;
            // XXX - float or [0,255] int ???
            System::Drawing::Imaging::ColorMatrix colorMatrix;
            colorMatrix.Matrix33 = (*it).second;
            alphaPaint.SetColorMatrix(%colorMatrix);

            canvas->DrawImage(*static_cast<msclr::gcroot<System::Drawing::Bitmap ^> *>(layerTile.bitmap.opaque),
                              System::Drawing::Rectangle(0, 0, layerTile.bitmap.width, layerTile.bitmap.height),
                              0, 0,
                              layerTile.bitmap.width,
                              layerTile.bitmap.height,
                              System::Drawing::GraphicsUnit::Pixel,
                              %alphaPaint);
        } finally {
            if (gotTile) {
                if (layerTile.bitmap.releaseData)
                    layerTile.bitmap.releaseData;
                if (layerTile.releaseData)
                    layerTile.releaseData(layerTile);
            }
        }
    }
    canvas->Flush();
    delete canvas;

    array<System::Byte> ^data = nullptr;

    System::IO::MemoryStream ^os = gcnew System::IO::MemoryStream();
    try {
        composite->Save(os, System::Drawing::Imaging::ImageFormat::Jpeg);
        data = os->ToArray();
    } finally {
        os->Close();
    }

    //return gcnew MobacMapTile(composite, data, expiration);
    return false;
#else
    return false;
#endif
}

void CustomMultiLayerMobacMapSource::setConfig(MobacMapSource::Config c)
{
    config = c;
}
